qt: Add PrivateSend tab in OptionsDialog, allow to show/hide PS UI

xdustinface opened the pull request:
- Move PrivateSend related settings from `Options -> Wallet` into a new tab `Options -> PrivateSend`
- Add `Enable PrivateSend interface` in `Options ->Wallet`. This checkbox allows to dynamically show/hide the PrivateSend related UI elements all over the place

Based on #3716 and #3715 

### Before
<img width="695" alt="Bildschirmfoto 2020-09-16 um 17 10 45" src="https://user-images.githubusercontent.com/35775977/93357004-232d7b00-f840-11ea-9812-7e89f2fc443c.png">

### With this PR

| PrivateSend UI hidden | PrivateSend options tab  |
| - | - |
| <img width="695" alt="Bildschirmfoto 2020-09-16 um 17 17 54" src="https://user-images.githubusercontent.com/35775977/93357573-c088af00-f840-11ea-8f8f-57edf1fa5131.png"> | <img width="695" alt="Bildschirmfoto 2020-09-16 um 17 18 01" src="https://user-images.githubusercontent.com/35775977/93357567-bff01880-f840-11ea-8c8a-83e19f8889ea.png"> |

UdjinM6 reviewed (commented):
Pls see suggestions below

UdjinM6 reviewed (commented):
Pls see below + there is a bug somewhere but I can't figure it out yet. The issue is like that: disable ps features in options, close and reopen the wallet (no ps ui as expected), enable ps in options -> still no ps ui on overview screen, ps tab and tx filters do appear though. If you close and reopen the wallet now ps ui would appear like it should.

xdustinface commented:
> there is a bug somewhere 

Ah right good catch! The PrivateSend UI update timer isn't created/started and the slots aren't connected if the wallet gets started with PrivateSend disabled. Fixing it 👍 

https://github.com/dashpay/dash/blob/develop/src/qt/overviewpage.cpp#L310
https://github.com/dashpay/dash/blob/develop/src/qt/overviewpage.cpp#L172

xdustinface commented:
@UdjinM6 Okaaay, there were even some more things i missed.. like two signal/slot connections weren't correct and whatnot.. see commits starting with 506614d. 

Now it instantly updates all UI elements if the checkbox gets toggled.. did only update the "PrivateSend" button of the options instantly before and then only all over the place after you pressed "Ok" in the dialog. It also recovers the previous state properly now if you reject the dialog.

PastaPastaPasta commented:
Why?? We just removed a basically pointless tab, since it made sense to have it under another tab, and yet now we're going to add another "PrivateSend" tab that makes total sense to be under the Wallet tab??

UdjinM6 commented:
> Why?? We just removed a basically pointless tab, since it made sense to have it under another tab, and yet now we're going to add another "PrivateSend" tab that makes total sense to be under the Wallet tab??

I would argue that Wallet tab has too many options already (7 checkboxes and 2 edit fields) and most of them are PS only, adding yet another one "Enable PrivateSend features" would make it look pretty clogged and weird imo. The new Wallet tab has 4 checkboxes and the new PrivateSend tab has 4 checkboxes and 2 edit fields, the later is exactly as for the new Main tab after #3709 on non-mac machines (on mac it's 1 checkbox and 2 edit fields only). If we ever think about exposing more PS options in GUI (and I think it would actually make sense to do) e.g. denoms goals/limits or number of mixing sessions etc. then having a separate tab is the only way to go imo.

PastaPastaPasta commented:
concept ACK on splitting settings out from inside of wallet. however privatesend must be enabled by default

UdjinM6 commented:
Needs rebase.

PastaPastaPasta commented on src/qt/optionsdialog.cpp:
```diff
@@ -179,6 +187,15 @@ void OptionsDialog::setModel(OptionsModel *_model)
+        if (strLabel.contains("-enableprivatesend")) {
+            bool fEnabled = CPrivateSendClientOptions::IsEnabled();
+            ui->privateSendEnabled->setChecked(fEnabled);
+            ui->privateSendEnabled->setEnabled(false);
```
I don't understand why we are setting it to false...

UdjinM6 replied:
Because cmd-line options override UI settings?

PastaPastaPasta replied:
But then why is it set to false if privatesend is enabled? I think that might be what I don't understand

xdustinface replied:
set to false == disable the checkbox

xdustinface replied:
and its not if PrivateSend enabled. Its only if "-enableprivatesend" is set on commandline to 0 or 1, then the checkbox gets disabled while it still represents the actual state.

PastaPastaPasta replied:
Hmm, okay, so probably makes sense to add a comment?
`
// If -enableprivatesend was passed in on the command line, set the "Enable PrivateSend" button to the value given via commandline and disable the checkbox, ie make it unclickable
`

Also, does this work if I use `-noenableprivatesend`?

xdustinface replied:
> Also, does this work if I use `-noenableprivatesend`?

Well, thats no valid commandline parameter unless just putting a `no` in front of a parameter is some hidden feature.. so not sure what should work then :D

PastaPastaPasta replied:
It's some hidden feature :D I use it often

xdustinface replied:
it is 😆 ok wait! About the comment, thats pretty clear actually, isnt it?

PastaPastaPasta replied:
I mean... I didn't find it super clear what was going on, but I am slow sometimes so :shrug: I'm leaning twords adding the comment

xdustinface replied:
Ok checking with the global args doesn't seem to work because it gets set in optionsmodel thats what i wanted to do but now just see 680e85bc3e6d121568cfb58916121f7337cae6ce

PastaPastaPasta replied:
Looks good to me :D

PastaPastaPasta commented on src/qt/optionsmodel.cpp:
```diff
@@ -113,6 +113,11 @@ void OptionsModel::Init(bool resetSettings)
         settings.setValue("digits", "2");
 
     // PrivateSend
+    if (!settings.contains("fPrivateSendEnabled"))
+        settings.setValue("fPrivateSendEnabled", true);
+    if (!gArgs.SoftSetBoolArg("-enableprivatesend", settings.value("fPrivateSendEnabled").toBool()))
+        addOverriddenOption("-enableprivatesend");
```
Please add brackets

xdustinface replied:
See 384dfb7

PastaPastaPasta reviewed (commented):
![cropped names](https://user-images.githubusercontent.com/6443210/94301368-f96c0680-ff59-11ea-9021-276614bca42d.png)
Adding the tab resulted in some not great cropping at my maybe default gui settings

PastaPastaPasta commented:
Starting with default settings made it way worse... 
![default](https://user-images.githubusercontent.com/6443210/94301503-30dab300-ff5a-11ea-86c4-e5bcd95a95ad.png)

PastaPastaPasta commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -234,25 +244,66 @@
+          <item>
+           <widget class="QCheckBox" name="privateSendEnabled">
+            <property name="toolTip">
+             <string>Show mixing interface on Overview screen and reveal PrivateSend screen which allows to spend fully mixed coins only. A new tab with more settings will also appear in this dialog, please make sure to check them before mixing your coins.</string>
```
This tooltip needs to wrap onto multiple lines

PastaPastaPasta commented:
Also, `noenableprivatesend` gets viewed by gui as `enableprivatesend=0` so we probably don't need that change specific to `noenableprivatesend`

xdustinface commented:
> Also, `noenableprivatesend` gets viewed by gui as `enableprivatesend=0` so we probably don't need that change specific to `noenableprivatesend`

Reverted and added a linebreak to the tooltip, should be good enough in two lines?

PastaPastaPasta commented on src/qt/optionsdialog.cpp:
```diff
@@ -190,6 +190,9 @@ void OptionsDialog::setModel(OptionsModel *_model)
         if (strLabel.contains("-enableprivatesend")) {
+        // If -enableprivatesend was passed in on the command line, set the checkbox
+        // to the value given via commandline and disable it (make it unclickable).
+        // Note: Check without "-" to make it work with the "no" prefix commandline feature.
```
```suggestion
```

UdjinM6 commented:
Re cropping: an easy fix would be to bump https://github.com/dashpay/dash/blob/develop/src/qt/res/css/general.css#L1257 - 635px worked ok-ish in my case but might need to bump it even more for Windows, smth like 700px or so I guess.

PastaPastaPasta commented:
> Re cropping: an easy fix would be to bump https://github.com/dashpay/dash/blob/develop/src/qt/res/css/general.css#L1257 - 635px worked ok-ish in my case but might need to bump it even more for Windows, smth like 700px or so I guess.

I had to go all the way up to 710px in order to make it work on ubuntu with default gui settings

UdjinM6 commented:
ESC key no longer cancels the changes that were made to ps checkbox, should do smth like 5ea108b88f

xdustinface commented:
@UdjinM6 good catch 👍 Added 06cdc89 instead which also takes care of other reject reasons like clicking the window's "x" close button.

UdjinM6 reviewed (approved):
ACK

(assuming font issues will be solved via #3734 )

PastaPastaPasta reviewed (approved):
ACK
